compiler/rustc_passes/src/dead.rs

```diff
@@ -787,22 +790,41 @@ impl<'tcx> Visitor<'tcx> for DeadVisitor<'tcx> {
         id: hir::HirId,
     ) {
         if self.should_warn_about_variant(&variant) {
-            self.warn_dead_code(variant.id, variant.span, variant.ident.name, "constructed");
+            self.warn_dead_code(variant.id, variant.span, variant.ident.name, "constructed", true);
         } else {
             intravisit::walk_variant(self, variant, g, id);
         }
     }
 
     fn visit_foreign_item(&mut self, fi: &'tcx hir::ForeignItem<'tcx>) {
         if self.should_warn_about_foreign_item(fi) {
-            self.warn_dead_code(fi.hir_id(), fi.span, fi.ident.name, "used");
+            self.warn_dead_code(fi.hir_id(), fi.span, fi.ident.name, "used", true);
         }
         intravisit::walk_foreign_item(self, fi);
     }
 
     fn visit_field_def(&mut self, field: &'tcx hir::FieldDef<'tcx>) {
         if self.should_warn_about_field(&field) {
-            self.warn_dead_code(field.hir_id, field.span, field.ident.name, "read");
+            debug!(?self.ignored_struct_def);
+            let struct_did = if let Some(parent_hir_id) = self.tcx.hir().find_parent_node(field.hir_id)
+                && let Some(parent_node) = self.tcx.hir().find(parent_hir_id)
+                && let Node::Item(hir::Item{kind: hir::ItemKind::Struct(..), ..}) = parent_node
+                && let Some(did) = self.tcx.hir().opt_local_def_id(parent_hir_id)
```

Comment:
I think you could move this logic to line 705, using the `id` parameter that already exists in that function. With that, you can get the hir parent (or maybe just get the hir owner), check if it's a struct, then do `ignored_struct_def.insert(did)`. That would avoid having to pass a `with_note: bool` parameter. Does that make sense?

Reply:
I didn't take that approach because I thought `warn_dead_code` should not decide if a note should be emitted or not inside it but just emit warnings. Should I fixed it?

Reply:
I think `warn_dead_code` already decides whether a note should be emitted or not, so modifying that condition further shouldn't be an issue. `warn_dead_code` also already takes `&mut self` too. :)

Reply:
Understood. I will fix it.